libqtile/command/base.py

```diff
@@ -55,11 +57,118 @@ class CommandException(Exception):
+        # be parsed here to check for new commands.
+        if getattr(cls, "_command_object", "") == cls.__name__:
+            return super(CommandObject, cls).__new__(cls)
+
+        commands = {}
+        cmd_s = set()
+
+        # We need to iterate over the class's inherited classes in reverse order
+        # We reverse the order so the exposed command will always be the latest
+        # definition of the method.
+        for c in reversed(list(cls.__mro__)):
+            base_classes = []
+
+            for method_name in list(c.__dict__.keys()):
+                method = getattr(c, method_name, None)
+
+                if method is None:
+                    continue
+
+                # Ignore methods that were defined by subclasses
+                for b in base_classes:
+                    if method == getattr(b, method_name, None):
+                        continue
+
+                # If the command has been exposed, add it to our dictionary
+                # If the method name is already in our dictionary then bind the
+                # latest definition to that command
+                if hasattr(method, "_cmd") or method_name in commands:
+                    commands[method_name] = method
+                # For now, we'll accept the old format `cmd_` naming scheme for
+                # exposing commands.
+                # NOTE: This will be deprecated in the future
+                elif method_name.startswith("cmd_"):
+                    cmd_s.add(method_name)
+                    commands[method_name[4:]] = method
+
+                # Expose additional names
+                for mapping in getattr(method, "_mapping", list()):
+                    setattr(cls, mapping, method)
+                    commands[mapping] = method
+
+                base_classes.append(method_name)
+
+        if cmd_s:
+            names = " ".join(cmd_s)
+            msg = (
+                f"The use of the 'cmd_' prefix to expose commands via IPC "
+                f"is deprecated. Methods should use the "
+                f"@CommandObject.expose() decorator instead. "
+                f"Please update: {names}"
+            )
+            logger.warning(f"Deprecation Warning: {msg}")
+
+        # Record the object as being parsed.
+        cls._command_object = cls.__name__
+
+        # Store list of exposed commands
+        cls._commands = commands
+
+        return super(CommandObject, cls).__new__(cls)
+
+    def expose(  # type: ignore
+        name: Optional[Union[str, List[str]]] = None  # noqa: N805
+    ) -> Callable:
+        """
+        Decorator to expose commands to IPC.
+
+        The exposed command will have the name of the defined method.
+
+        Methods can also be exposed via multiple names by passing the names of
+        the additional method.
+
+        e.g. if a layout wants "up" and "previous" to call the
+        same method:
+
+        @CommandObject.expose("previous")
```

Comment:
If people accidentally use `@expose` rather than `@expose()`, this may cause some odd errors. It should be possible with some introspection to determine if we are in the former case. We could allow that, and just adjust the logic to set the right values on the function, or we should try to raise an error.

Reply:
Nice idea.